### ISSUE TITLE
SqlTraceFlag: New resource proposal

### ISSUE THREAD

**Issue by Zuldan** [dsccommunity/SqlServerDsc#110]
Example code based on https://meltondba.wordpress.com/2015/12/02/modifying-sql-server-startup-parameters/

```
$server = 'MyServer'
$sqlservice = "MSSQLSERVER"
$sqlagentservice = "SQLSERVERAGENT"
$flagsToAdd = ';-T1117;-T1118;-T3226'

Add-Type -AssemblyName "Microsoft.SqlServer.SqlWmiManagement,Version=11.0.0.0,Culture=neutral,PublicKeyToken=89845dcd8080cc91"
$sqlwmi = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $server
$wmisvc = $sqlwmi.Services | where {$_.name -eq $sqlservice}
$wmisvc.StartupParameters = $wmisvc.StartupParameters + $flagsToAdd
$wmisvc.Alter()

$wmisvc.Stop()
Start-Sleep -seconds 15
$wmisvc.Start()

$wmiAgent = $sqlwmi.Services | where {$_.name -eq $sqlagentservice}
$wmiAgent.Start()
```

**Comment by Zuldan** [comment]
So here's a quick Script I made to temporarily do the job.

```
Script SQLTraceFlags {
    DependsOn = '[xSQLServerSetup]SQLServerSetup'
    GetScript = {
        $Status = $True
        Add-Type -AssemblyName "Microsoft.SqlServer.SqlWmiManagement,Version=$($Using:Node.SQLVersion).0.0.0,Culture=neutral,PublicKeyToken=89845dcd8080cc91"
        $SQLManagement = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $Using:Node.NodeName
        $WMIService = $SQLManagement.Services | Where-Object -FilterScript { $PSItem.Name -eq $Using:Node.SQLInstanceName }

        $TraceFlagList = ($Using:Node.SQLTraceFlags).Split(';')
        [System.Collections.ArrayList]$ParameterList = $WMIService.StartupParameters.Split(';')

        # Test for extra flags
        foreach ($Parameter in $WMIService.StartupParameters.Split(';')) {
            if ($Parameter -like '-T*' -and $Parameter -notin $TraceFlagList) {
                $Status = $False
            }
        }

        # Test for missing flags
        foreach ($Flag in $TraceFlagList) {
            if ($Flag -notin $ParameterList) {
                $Status = $False
            }
        }

        @{
            GetScript = $GetScript
            SetScript = $SetScript
            TestScript = $TestScript
            Result = $Status
        }
    }
            
    SetScript = {
        Add-Type -AssemblyName "Microsoft.SqlServer.SqlWmiManagement,Version=$($Using:Node.SQLVersion).0.0.0,Culture=neutral,PublicKeyToken=89845dcd8080cc91"
        $SQLManagement = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $Using:Node.NodeName
        $WMIService = $SQLManagement.Services | Where-Object -FilterScript { $PSItem.Name -eq $Using:Node.SQLInstanceName }

        $TraceFlagList = ($Using:Node.SQLTraceFlags).Split(';')
        [System.Collections.ArrayList]$ParameterList = $WMIService.StartupParameters.Split(';')

        # Removing extra flags
        foreach ($Parameter in $WMIService.StartupParameters.Split(';')) {
            if ($Parameter -like '-T*' -and $Parameter -notin $TraceFlagList) {
                $ParameterList.Remove($Parameter) | Out-Null
            }
        }

        # Add missing flags
        foreach ($Flag in $TraceFlagList) {
            if ($Flag -notin $ParameterList) {
                $ParameterList.Add($Flag) | Out-Null
            }
        }

        $WMIService.StartupParameters = $ParameterList -join ';'
        $WMIService.Alter()

        $WMIService.Stop()
        Start-Sleep -Seconds 10
        $WMIService.Start()

        ($SQLManagement.Services | Where-Object -FilterScript { $PSItem.Name -eq 'SQLSERVERAGENT' }).Start()
    }

    TestScript = {
        $Status = $True
        Add-Type -AssemblyName "Microsoft.SqlServer.SqlWmiManagement,Version=$($Using:Node.SQLVersion).0.0.0,Culture=neutral,PublicKeyToken=89845dcd8080cc91"
        $SQLManagement = New-Object Microsoft.SqlServer.Management.Smo.Wmi.ManagedComputer $Using:Node.NodeName
        $WMIService = $SQLManagement.Services | Where-Object -FilterScript { $PSItem.Name -eq $Using:Node.SQLInstanceName }

        $TraceFlagList = ($Using:Node.SQLTraceFlags).Split(';')
        [System.Collections.ArrayList]$ParameterList = $WMIService.StartupParameters.Split(';')

        # Test for extra flags
        foreach ($Parameter in $WMIService.StartupParameters.Split(';')) {
            if ($Parameter -like '-T*' -and $Parameter -notin $TraceFlagList) {
                $Status = $False
            }
        }

        # Test for missing flags
        foreach ($Flag in $TraceFlagList) {
            if ($Flag -notin $ParameterList) {
                $Status = $False
            }
        }

        $Status -eq $True
    }
}
```

**Comment by ArieHein** [comment]
@Zuldan trace flags can be changed via a sql script, so you can use the xSQLServerScript resource.

IIRC they appear in the registry as well, so obviously the Registry resource can be used.

not sure how this will be used if the trace flag requires the service be restarted

**Comment by randomnote1** [comment]
I concur that this would be a good resource to add. Should the proposed resource name be "xSQLServerTraceFlag" to conform to naming conventions?

**Comment by johlju** [comment]
'xSQLServerTraceFlag' would be correct, yes. 

But a question. Trace flags are set as part of the startup options, right? So maybe the resource should be namned 'xSQLServerStartupOptions' if there are other options that could be implemented besides the trace flags. Trace flags could be a parameter with a array of integers.

**Comment by randomnote1** [comment]
Startup options could be interesting. Here's official article for reference:

https://msdn.microsoft.com/en-us/library/ms162819.aspx

**Comment by johlju** [comment]
The name should be 'xSQLServerStartupOption' (singular). I wrote it in plural above. :)
Suggested schema for the an initial resource. I could be improved to support more parameters in the future. Would this work?

```
[ClassVersion("1.0.0.0"), FriendlyName("xSQLServerStartupOption")]
class MSFT_xSQLServerStartupOption : OMI_BaseResource
{
    [Write, Description(""), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Key, Description("")] String SQLServer;
    [Key, Description("")] String SQLInstanceName;
    [Write, Description("An array of trace flags.")] Uint32[] TraceFlag;
    [Write, Description("Forces a reboot after the desired state is set. Default values is $false.")] Boolean ForceReboot;
};
```

**Comment by randomnote1** [comment]
I'm not a fan of the "ForceReboot" verbiage because it implies the Windows server will be rebooted. I think "RestartSqlService" or "RestartInstance" would more accurately portray what needs to occur.

Were you thinking of adding one parameter for each argument sqlservr.exe supports?

**Comment by emitodiaz** [comment]
Hi @randomnote1 
Actually I reboot sqlserver and later the windows server trying to resolve this error but with no luck.
I don't understand what do you mean by adding a parameter to sqlservr.exe, could you explain plis.
Thanks.

**Comment by randomnote1** [comment]
@emitodiaz, I think we've crossed threads here. I'm commenting on issue #110.

**Comment by emitodiaz** [comment]
@randomnote1 yes, I'm sorry....

**Comment by johlju** [comment]
My bad. Meant for it to be `ForceRestart`. But you are right, using word 'force' are implied that sometimes it is started by itself, which it is not. And restart of what? :) 'RestartSqlService' doesn't sound good either, because if SQL Agent is started that need to be restarted as well. So `RestartInstance` sound like the best option so far.

I'm happy for it to include all parameters as long as there is code behind to support the parameters.

**Comment by randomnote1** [comment]
The other option we could use is instead of "TraceFlag" we could accept an array or hash table in "StartupParameter". This would allow any of the parameters to be used with minimal code on the back end. The down side to this approach could be that all of the default startup parameters would need to be specified in the DSC when additional startup parameters are specified.

**Comment by johlju** [comment]
If we use the hash table you suggest, the user needs to know exactly to format each parameter. And the user cannot use the same configuration for different versions of SQL Server (if it would change drastically, like it might do for SQL Server for Linux). If we have a parameter for each option, like trace flag, the user can add or remove just one trace flag without knowing the rest of the startup options in the string (as you mention above).
Hypothetical if an update/upgrade of SQL Server changes the startup options in any way, the config will just remove that and might make SQL Server in an unstable state.

But that said, I'm open to use a hash table as you suggest, it's better than not having the resource at all. As long as we don't do work that will be obsolete down the road. :smile:

**Comment by Zuldan** [comment]
Hi all, I've built a fully functional DSC class based resource that we now have in production. I guess someone could convert it to MOF based DSC resource and use it as a starting point? I wrote it using the style guidelines (as much as possible) https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md

If anyone is interested I could post the code on GitHub.

**Comment by johlju** [comment]
We are allowed to have class based resources as well. 
If you don't want to PR in the resources yourself, and instead want to share the code and allow us to re-use it, then it absolutely would be interesting to see the code. 
Thank you.

**Comment by Zuldan** [comment]
Wow I didn't think class resources were allowed yet. I know it was a discussion point a little while ago. I would love to do PR (although never done one before) but it doesn't have any unit testing. I'll post the code on GitHub just before or just after Xmas.

**Comment by johlju** [comment]
We do need unit tests for the resource before I can merge it. This would be the first class based resource here, so might be some issues along the way. 
If you want to PR in your resource we will work together to move it along :)

Could you post the schema (parameters) here as well when you have time? So that we could discuss that in regards to what has been discussed here earlier.

**Comment by Zuldan** [comment]
I'll post the parameters later today and post the full code on Gist. Many thank for your offer of help. If you think it's worth getting into GitHub / doing a PR then I'll work on that (with your help).

**Comment by Zuldan** [comment]
Here it just https://github.com/Zuldan/cSQLServerTraceFlag. Decided just to upload the whole thing. Threw together some examples (haven't tested them so they probably don't work yet). 

Parameters as follows;

```
        cSQLServerTraceFlag SQLTraceFlag
        {
            Ensure            = 'Present'
            SQLVersion        = 13
            SQLInstanceName   = 'MSSQLSERVER'
            TraceFlag         = 'T834','T1117','T1118','T2371','T3226'
            RestartSQLService = $True
        }
```

I was originally going to write a resource to modify startup options but then I figured those should probably be configured in xSQLServerSetup (property for this yet to exist of course) as startup options are usually configured as a once off where as you may want to change trace flags from time to time (hence a dedicated resource). Also, Microsoft doesn't group together start up options with trace flags on MSDN. Trace flags has it's own page, https://msdn.microsoft.com/en-us/library/ms188396.aspx as they can be configured by using start up options or within the SQL server, EG DBCC TRACEON (2528, -1) .

Anyway, let the discussion continue. I'm just happy to contribute in anyway no matter what everyone decides on.

**Comment by johlju** [comment]
@Zuldan you have the code already finished for setting the trace flag. We just need to incorporate it into this module, using the helper functions, fixing tests, and so on. It will be an undertaking, but hopefully it will be fun. 😄  

I'm against setting DBE startup options with the xSQLServerSetup resource, unless there is an option for that when calling setup.exe. Also think that xSQLServerSetup resource is complex enough already. I can be swayed if others are for that approach.
But changing the startup options isn't a regular thing, so I still think a separate resource for this is the best option. And since the the startup options are for the Database Engine, the resource name should reflect that. I see that the resource could potentially have the parameters outline in the following article; https://msdn.microsoft.com/en-us/library/ms190737.aspx. But we could start with just the TraceFlag.

We have two suggestions (as schema.mof, but can be class based). And I want to throw in a third suggestion too. :)
Do we have more suggestions?

**Suggestion 1**
A string that sets all options. User needs to know what options are set already.

```
[ClassVersion("1.0.0.0"), FriendlyName("xSQLServerDatabaseEngineStartupOption")]
class MSFT_xSQLServerDatabaseEngineStartupOption : OMI_BaseResource
{
    [Write, Description(""), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Key, Description("")] String SQLServer;
    [Key, Description("")] String SQLInstanceName;
    [Write, Description("A string with all the startup options to replace all existing options")] String StartupParameter;
    [Write, Description("Forces a restart of the Database Engine service and dependent services after the desired state is set. Default values is $false.")] Boolean RestartInstance;
};
```

**Suggestion 2**
A parameter for each option to set. The values in the parameter will be added to or removed from the existing values.

```
[ClassVersion("1.0.0.0"), FriendlyName("xSQLServerDatabaseEngineStartupOption")]
class MSFT_xSQLServerDatabaseEngineStartupOption : OMI_BaseResource
{
    [Write, Description(""), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Key, Description("")] String SQLServer;
    [Key, Description("")] String SQLInstanceName;
    [Write, Description("An array of trace flags to be added or removed from the startup options")] Uint32[] TraceFlag;
    [Write, Description("Forces a restart of the Database Engine service and dependent services after the desired state is set. Default values is $false.")] Boolean RestartInstance;
};
```

**Suggestion 3**
Three parameters for each option to set. One which replaces all existing values, a second that only adds values to the existing values, and a third that only removes values from the existing values.

```
[ClassVersion("1.0.0.0"), FriendlyName("xSQLServerDatabaseEngineStartupOption")]
class MSFT_xSQLServerDatabaseEngineStartupOption : OMI_BaseResource
{
    [Write, Description(""), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Key, Description("")] String SQLServer;
    [Key, Description("")] String SQLInstanceName;
    [Write, Description("An array of trace flags that startup options should have")] Uint32[] TraceFlag;
    [Write, Description("An array of trace flags to be added to the existing trace flags")] Uint32[] TraceFlagToInclude;
    [Write, Description("An array of trace flags to be removed from the existing trace flags")] Uint32[] TraceFlagToExclude;
    [Write, Description("Forces a restart of the Database Engine service and dependent services after the desired state is set. Default values is $false.")] Boolean RestartInstance;
};
```

**Comment by johlju** [comment]
Sorry, we had a forth suggestion. The initial suggestion, and the same one that @Zuldan's class based resource.
The name 'xSQLServerTraceFlag' I think is to generic, it should be either 'xSQLServerDatabaseEngineTraceFlag' or 'xSQLServerDatabaseEngineStartupOptionTraceFlag'. For now I used 'xSQLServerDatabaseEngineTraceFlag'.

**Suggestion 4**
A resource for setting just the trace flags in the start options.

```
[ClassVersion("1.0.0.0"), FriendlyName("xSQLServerDatabaseEngineTraceFlag")]
class MSFT_xSQLServerDatabaseEngineTraceFlag : OMI_BaseResource
{
    [Write, Description(""), ValueMap{"Present","Absent"}, Values{"Present","Absent"}] String Ensure;
    [Key, Description("")] String SQLServer;
    [Key, Description("")] String SQLInstanceName;
    [Write, Description("An array of trace flags to be added or removed from the startup options")] Uint32[] TraceFlag;
    [Write, Description("Forces a restart of the Database Engine service and dependent services after the desired state is set. Default values is $false.")] Boolean RestartInstance;
};
```

**Comment by arcotek-ltd** [comment]
Thank you to @Zuldan for the xSQLTraceFlag resource. I couldn't wait for the testing and commit to SqlServerDsc, so I've gone ahead and added it to my copy of SqlServerDsc. (Bad, I know). I've always wondered how to have class and mof based resources in the same module. Obvious to you guys, I'm sure, but I've managed to work it out. Well, `Get-DSCResource -Module SqlServerDsc` picks it up. Not sure if it's just trashed everything else.

Anyway, what I did was:

1.  Download [cSqlServerTraceFlag](https://github.com/Zuldan/cSQLServerTraceFlag/archive/master.zip).
2.  Extract ZIP and rename to cSqlServerTraceFlag
3.  Drop the folder (cSqlServerTraceFlag) into "%ProgramFiles%\WindowsPowerShell\Modules\SqlServerDsc\DSCResources"
4.  Edit .\Modules\SqlServerDsc\SQLServerDsc.psd1, adding:
   a.  `NestedModules = @('DSCResources\cSQLServerTraceFlag\cSQLServerTraceFlag.psm1')`
   b.  `DscResourcesToExport = '*'`

Now to test...
HTH

**Comment by johnmart82** [comment]
I'm looking at using DSC for my SQL Server deployment and management within my organization, one of the things that I am looking at is needing to ensure certain trace flags are set on the server. What is the current state of plat with adding this reource to the SqlServerDsc module and is there anything that I can do to help at all?

For reference the trace flags I will be looking to set by default are as follows;
- 460 [Data Truncation Message]
- 3226 [Stop logging successful backup messages]
- 3625 [Masking error messages from non-sysadmins]
- 9567 [Compression on seeding log traffic]

Plan is to set these at deployment time and then monitor and enforce them via DSC.

I like the approach of sticking with trace flags only in the resource as it is a clearly defined set of configuration options, which I feel are distinct from the other startup options.

**Comment by johlju** [comment]
@johnmart82 Nobody have started work on this as far as I know - if you want to help put and create a new resource for this, then you are more than welcome. I will be happy to review a PR for this.

**Comment by mdaniou** [comment]
I would happily welcome that resource too ! 👍 :)

**Comment by ArieHein** [comment]
As mentioned way up the history, startup settings and trace flags can be set via the registry, so using the registry is one way without waiting for the resource to offer it.
Here is a related information about powershell and SQL statup settings and flags:
https://blog.waynesheffield.com/wayne/archive/2017/09/registry-sql-server-startup-parameters/

Do think if you really need this to be set via DSC. Remember DSC should be used to create a "base" configuration and enforcing it. So say you have 10 SQL servers all created from same configuration and MOF and all have same settings and configs and flags, now you have an issue with one server and you need to change its tracing flag to overcome an issue...what happens 30 min later ?  flag changes...but service isn't restarted so that flag isn't in effect..do you then disable the agent, use more then one configuration ith different flags to accommodate ? send the new confirmation just to one server ?

I would dare say that if you have settings like location of master, tempdb, that can be in a configuration but maybe not the trace flags as those might change on a more "often" basis compare to folder locations. AS long as you remember that you still need to restart the service to have the new settings in effect which isn't the best way to control via DSC for that matter.

**Comment by mdaniou** [comment]
@Zuldan @johlju 
I had a look at the code and I am interested in using class eventually.
However I wonder what is the real benefit of it ...
Should we tend to it ?
I am more a devops than a developer if you see what I mean.

**Comment by johlju** [comment]
It has been attempt to add a class-bases resource to this module before. The problem then was how to add more than one class-based resource and on the same time use the mof-resource. There was also a problem with testing. You had to run the unit test in a new session for each run when coding the tests. There might work if using the same approach as this repo is using; https://github.com/bgelens/DockerDsc.
Would it benefit to use class-based resource? I would say not, not until there is better support for classes and we can run the resources in PowerShell Core.

**Comment by mdaniou** [comment]
Ok thanks for the answer

**Comment by craig-martin** [comment]
Regarding multiple class-based resources I have it working in an [isolated test module](https://github.com/craig-martin/TestModule).  Considering converting some of mof-based DSC resources from a module I'm working on and just getting started with class-based resources.  Spent a couple hours struggling with differences in how modules are loaded, but got it solved.  

The lack of people using them makes me a bit nervous, but I really like using classes to design DSC schema instead of mof, because some of my resources have nested objects which gets weird in mof.

**Comment by johlju** [comment]
That’s great work, it proves the theory I have been thinking about! I’m really looking forward to when it’s possible to run unit tests several times in a row in the same session. That is really what is preventing me from going with class-based resources. 
It do work in latest PS Core, I think. But might not yet in the latest Windows PowerShell. Although it was a while since I tested it.

**Comment by craig-martin** [comment]
I wish classes could easily be instantiated and tested with New-Object; seems weird that you can't use New-Object to create classes defined in a module, but I get that it is by design. Using Invoke-DscResource seems to work fine. Here's $PSVersionTable from my machine where I'm testing:
```

Name                           Value                                                                                                                                                                               
----                           -----                                                                                                                                                                               
PSVersion                      5.1.17763.316                                                                                                                                                                       
PSEdition                      Desktop                                                                                                                                                                             
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                             
BuildVersion                   10.0.17763.316                                                                                                                                                                      
CLRVersion                     4.0.30319.42000                                                                                                                                                                     
WSManStackVersion              3.0                                                                                                                                                                                 
PSRemotingProtocolVersion      2.3                                                                                                                                                                                 
SerializationVersion           1.1.0.1                                                                                                                                                                             
```

Going to keep using MOF-based DSC resources for now but might circle back once my module is done and switch to class-based DSC resources now that I know they work well enough.

**Comment by johlju** [comment]
Running Invoke-DscResources does an integration test, and actually does the change. What I mean was a unit test that mocks the calls, to test all the code path and different scenarios but does not actually change anything on a target node. Unit test should be able to run on a dev machine without any SQL Server it modules installed. 
It particular unit test that can’t be (or couldn’t be) run more than once in the same session.

**Comment by craig-martin** [comment]
Got it, I've just been running integration tests, by calling the resource modules directly outside the LCM, and by using a single-item configuration in the LCM.  Going to have to take a closer look at your unit tests to see what I can borrow ;-)

**Comment by joshshand** [comment]
This resource request can probably now be closed with the release of 15.1.1. SqlTraceFlag is now available and tested working!

**Comment by johlju** [comment]
@joshshand Thank you.
